File: Timer/src/main/java/org/elkoserver/foundation/timer/TimerQEntry.java
package org.elkoserver.foundation.timer;

import java.time.Clock;

/**
 * An entry in the timer event queue.
 */
class TimerQEntry implements Comparable<TimerQEntry>
{
    boolean myRepeat;
    long myDelta;
    long myWhen;
    TimerWatcher myTarget;
    TimerQEntry myNext;

    TimerQEntry(boolean repeat, long delta, TimerWatcher target, Clock clock) {
        myRepeat = repeat;
        myDelta = delta;
        myWhen = clock.millis() + delta;
        myTarget = target;
        myNext = null;
    }

    public int compareTo(TimerQEntry other) {
        if (other != null) {
            return Long.compare(myWhen, other.myWhen);
        } else {
            throw new ClassCastException();
        }
    }

    public boolean equals(Object obj) {
        if (obj instanceof TimerQEntry) {
            TimerQEntry other = (TimerQEntry) obj;
            return myWhen == other.myWhen;
        } else {
            return false;
        }
    }
}

File: Timer/src/main/kotlin/org/elkoserver/foundation/timer/TimerQEntry.kt
package org.elkoserver.foundation.timer

import java.time.Clock

/**
 * An entry in the timer event queue.
 */
internal class TimerQEntry(internal var myRepeat: Boolean, internal var myDelta: Long, var myTarget: TimerWatcher, clock: Clock) : Comparable<TimerQEntry?> {
    @JvmField
    var myWhen = clock.millis() + myDelta

    @JvmField
    var myNext: TimerQEntry? = null

    override fun compareTo(other: TimerQEntry?) =
            if (other != null) {
                myWhen.compareTo(other.myWhen)
            } else {
                throw ClassCastException()
            }

    override fun equals(other: Any?): Boolean {
        return if (other is TimerQEntry) {
            myWhen == other.myWhen
        } else {
            false
        }
    }

    override fun hashCode(): Int {
        return myWhen.hashCode()
    }
}
